### DISCUSSION TITLE
MQTT not authorized

### DISCUSSION THREAD

**Discussion by tp-x12** [tbnobody/OpenDTU#1403] [Q&A]
Hello,

I'm trying to connect my OpenDTU (v23.9.18 on an ESP32) to my MQTT-Broker. The Broker is running on an Ubuntu 20.04. It requires authentification and TLS. There are 8 other clients which are able to connect successfully to this broker. 

However, my OpenDTU is not able to authentificate (TLS is working, though):
`> Disconnect reason:MQTT_NOT_AUTHORIZED`

In the logfile of the broker, I'm seeing a corresponding CONNACK return code
```
1696278412: New connection from 192.168.***.24 on port 8883.
1696278413: Sending CONNACK to 192.168.***.24 (0, 5)
1696278413: Socket error on client <unknown>, disconnecting.
```

To my knowledge, this typically points towards a wrong user / password combination. However, this combination is correct. The MQTT access-config includes 
`OpenDTU:abcde`

... and the openDTU json has 
`"mqtt":{"enabled":true,"hostname":"192.168.***.30","port":8883,"username":"OpenDTU","password":"abcde","topic":"solar/","retain":true,"publish_interval":5,"clean_session":true,"lwt":{"topic":"dtu/status","value_online":"online","value_offline":"offline"},"tls":{"enabled":true,"root_ca_cert":"-----BEGIN `

Using these credentials, I'am able to connect to the broker via mosquitto_sub on Windows
`mosquitto_sub -h 192.168.***.30 -p 8883 -t # -v --cafile C:\***\Arduino\certs\ca.crt --cert C:\***\Arduino\certs\***.crt --key C:\Users\tp\Documents\Arduino\certs\***.key -u OpenDTU -P abcde`

With OpenDTU, I'm not able to authentificate with these credentials or with any other credentials which are known to work with other clients.

Has anybody got any idea how to further debug this? Thanks for your efforts!

**Comment by avandorp** [comment]
The ESP32 is somewhat limited in its ability to perform TLS. Do you have an exceptionally large key? I use MQTT with TLS and authentication against a Mosquitto installation with letsEncrypt certificates, that works very well so far.

**Reply by tp-x12** [reply]
The length is standard, 2048 bits. I can use this type of setup with other ESP32 and ESP8266 without any problems. Thus, I do not believe that this is the cause. Thanks for the idea!

**Comment by tbnobody** [comment]
Please doublcheck that the server name (MQTT Broker) which you are entering in the MQTT settings of OpenDTU matches the CN of your server certificate. Otherwise the connection will fail.

**Reply by tp-x12** [reply]
Indeed, the CN name of the server certificate does not match the broker name. I will check this out, but it will take some time to reset the whole infrastructure. I'll let you know if that helps. Thanks for the idea!

**Reply by tp-x12** [reply]
Okay, tested a bit today. Recreated the certificates. Triple checked that the CN name in the certificate of the mosquitto broker matches the CN name in the client certificate. Both are the local ip adress of the of my server. Updatet OpenDTU with the new certificates. Still not working.

I'm sure that it is not a SSL error. If I add a random char to one of the certificates in OpenDTU, thus invalidating them, the console produces a different error: 
```
23:05:57.856 > Connecting to MQTT...
23:05:57.921 > Disconnected from MQTT.
23:05:57.921 > Disconnect reason:TCP_DISCONNECTED
```
If I remove that random char, the error changes back to 
```
Connecting to MQTT...
23:07:04.357 > Disconnected from MQTT.
23:07:04.357 > Disconnect reason:MQTT_NOT_AUTHORIZED
```
Additionally, if it was a SSL error, the mqtt broker would not be able to send a CONNACK (0,5) to the client, because it would not even understand the request. Then, I am also able to connect to the mqtt broker with this exact set of certificates and credentials from other clients. It is just the OpenDTU client that is not able to authenticate. And finally, if I allow clients to connect anonymously in the mqtt conf, the OpenDTU client works flawlessly. Thus, the issue seems to be that the OpenDTU client is not able to communicate the username and/or password in the right way to the broker. Would be happy if someone has any further ideas on how to debug this issue. And, thanks for your efforts so far!

**Reply by tbnobody** [reply]
What is the exact CN in your certificate and what are you entering in the host text field of opendtu?
As far as I know, only hostnames (not ips) are working.
Did you also paste your CA certificate into the text field?

**Reply by tp-x12** [reply]
it is the local ip of my server in both certificates (for the broker and the client), i.e. 192.168.209.30. The CA certificate is pasted into OpenDTU, too. But as I said above, I'm sure that it is not a problem with SSL. Even with SSL completely turned off and user management turned on in mosquitto conf, OpenDTU is not able to connect to the broker (CONNACK(0,5)) There seems to be a problem in communicating plaintext username and/or password to MQTT broker. Tried today all credentials that are in the access conf of MQTT, none is working. Maybe there is a verbose mode for the console to get further infos?

**Reply by tp-x12** [reply]
Just a small update: I followed your advise and changed the CN in the certificates to the hostname instead of the ip. Still getting CONNACK(0,5)

p.s.: But you're right, using ips instead of hostnames is deprecated (but still works). Mosquitto logs this warning at startup
